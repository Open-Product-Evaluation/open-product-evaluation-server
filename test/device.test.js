const users = require('../seeds/data/user/user')
const devices = require('../seeds/data/device/device')
const contexts = require('../seeds/data/context/context')
const { seedDatabase } = require('mongo-seeding')
const config = require('../config')
const request = require('./requesthelper')
const { getSeedID } = require('./helpers')

/* Functions for Querys */

function createDeviceQuery(name) {
  return {
    query: `mutation {
        createDevice(data: {name: "TestDevice"}) {
          device {
            name
            context {
              id
            }
          }
          token
        }
      }`,
  }
}

function updateDeviceQuery(deviceID, deviceName, context, owners) {
  return {
    query: `mutation {
      updateDevice(deviceID:"${deviceID}", data:{name:"${deviceName}",context:"${context}", owners: ${JSON.stringify(owners)} }){
        device{
          id
          name
          context{
            id
          }
          owners{
            id
          }
        }
      }
    }`,
  }
}

function deleteDeviceQuery(deviceID) {
  return {
    query: `mutation {
      deleteDevice(deviceID:"${deviceID}"){
        success
      }
    }`,
  }
}

function devicesQuery() {
  return {
    query: `{
      devices {
        id
        name
        context {
          id
        }
        owners {
          id
        }
      }
    }`,
  }
}

function deviceQuery(deviceID) {
  return {
    query: `{
      device(deviceID: "${deviceID}") {
        id
        name
        context {
          id
        }
        owners{
          id
        }
      }
    }`,
  }
}

/* Tests */

describe('Device', () => {
  describe('Anonym', async () => {
    beforeAll(() => seedDatabase(config.seeder))
    it('should create a Device [Mutation]', async () => {
      const query = createDeviceQuery('TestDevie')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data.createDevice.device).toMatchSnapshot()
      expect(data.createDevice.token.length).toBeGreaterThan(0)
      expect(errors).toBeUndefined()
    })
    it.skip('should deny creation of existing Device [Mutation]', async () => {
      const query = createDeviceQuery('TestDevie')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
  describe('Admin', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[1]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should return all devices [Query]', async () => {
      const query = devicesQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return device owned by User [Query]', async () => {
      const device = devices[3]
      const query = deviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return device not owned by User [Query]', async () => {
      const device = devices[1]
      const query = deviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should create device [Mutation]', async () => {
      const query = createDeviceQuery('TestDevice')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data.createDevice.device).toMatchSnapshot()
      expect(data.createDevice.token.length).toBeGreaterThan(0)
      expect(errors).toBeUndefined()
    })
    it('should update device owned by User [Mutation]', async () => {
      const device = devices[3]
      const context = contexts[0]
      const user = users[0]
      const query = updateDeviceQuery(getSeedID(device), 'RenamedTestDevice', getSeedID(context), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should update device not owned by User [Mutation]', async () => {
      const device = devices[1]
      const context = contexts[0]
      const user = users[0]
      const query = updateDeviceQuery(getSeedID(device), 'RenamedTestDevice', getSeedID(context), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should delete device owned by user [Mutation]', async () => {
      const device = devices[3]
      const query = deleteDeviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteDevice.success).toBe(true)
    })
    it('should delete device not owned by user [Mutation]', async () => {
      const device = devices[1]
      const query = deleteDeviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteDevice.success).toBe(true)
    })
  })
  describe('User', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[0]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should return all devices [Query]', async () => {
      const query = devicesQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return device owned by User [Query]', async () => {
      const device = devices[0]
      const query = deviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should not return device not owned by User [Query]', async () => {
      const device = devices[3]
      const query = deviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should create device [Mutation]', async () => {
      const query = createDeviceQuery('TestDevice')
      const res = await request.anon(query)
      const { data, errors } = res
      expect(data.createDevice.device).toMatchSnapshot()
      expect(data.createDevice.token.length).toBeGreaterThan(0)
      expect(errors).toBeUndefined()
    })
    it('should update device owned by User [Mutation]', async () => {
      const device = devices[0]
      const context = contexts[0]
      const user = users[0]
      const query = updateDeviceQuery(getSeedID(device), 'RenamedTestDevice', getSeedID(context), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should not update device not owned by User [Mutation]', async () => {
      const device = devices[3]
      const context = contexts[0]
      const user = users[0]
      const query = updateDeviceQuery(getSeedID(device), 'RenamedTestDevice', getSeedID(context), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should delete device owned by user [Mutation]', async () => {
      const device = devices[0]
      const query = deleteDeviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteDevice.success).toBe(true)
    })
    it('should not delete device not owned by user [Mutation]', async () => {
      const device = devices[3]
      const query = deleteDeviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
  describe('Device', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const query = {
        query: `mutation{createDevice(data:{name:"TestDevice"}){
          token
       }}`,
      }
      const { data, errors } = await request.anon(query)
      data.createDevice.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { createDevice: { token } } = data
      jwtToken = token
    })
    it('should not return all devices [Query]', async () => {
      const query = devicesQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should not return device owned by User [Query]', async () => {
      const device = devices[0]
      const query = deviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should update device [Mutation]', async () => {
      const device = devices[0]
      const context = contexts[0]
      const user = users[0]
      const query = updateDeviceQuery(getSeedID(device), 'RenamedTestDevice', getSeedID(context), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should not delete device [Mutation]', async () => {
      const device = devices[3]
      const query = deleteDeviceQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
})
