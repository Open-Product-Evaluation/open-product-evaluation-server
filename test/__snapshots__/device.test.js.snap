// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Device Admin schould create device [Mutation] 1`] = `
Object {
  "context": null,
  "name": "TestDevice",
}
`;

exports[`Device Admin schould return all devices [Query] 1`] = `
Object {
  "devices": Array [
    Object {
      "context": Object {
        "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
      },
      "id": "13624f62a9b2a01b18d0ae062bf9ad978d0fe5265b9e059957f15c331ae11282",
      "name": "Forum Fernseher",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
    Object {
      "context": Object {
        "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
      },
      "id": "9db9921b0d8132795cb5e77a90572fce5e94b90a8f44bd48417dff8cf7b66834",
      "name": "Forum Tablet",
      "owners": null,
    },
    Object {
      "context": Object {
        "id": "16613d16bf18d5674ad8c5d3ef071a3835ceb2b09fc3d5df592c0925f2e881f9",
      },
      "id": "3f4cd071e7f98511b24b448681003435e67b010c0fd271ebd494ac6537941088",
      "name": "Mensa Kiosk",
      "owners": null,
    },
    Object {
      "context": Object {
        "id": "16613d16bf18d5674ad8c5d3ef071a3835ceb2b09fc3d5df592c0925f2e881f9",
      },
      "id": "4ed5ac9fcf3638b1ef16c7800a3f6d7c5b4aa092b9afe1dca139c65e97f023f1",
      "name": "Mensa Tablet",
      "owners": Array [
        Object {
          "id": "bb9b812794666578a86254afca49deef281fd5caf4566db9915cf4d5ecde53dd",
        },
      ],
    },
  ],
}
`;

exports[`Device Admin schould return device not owned by User [Query] 1`] = `
Object {
  "device": Object {
    "context": Object {
      "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
    },
    "id": "9db9921b0d8132795cb5e77a90572fce5e94b90a8f44bd48417dff8cf7b66834",
    "name": "Forum Tablet",
    "owners": null,
  },
}
`;

exports[`Device Admin schould return device owned by User [Query] 1`] = `
Object {
  "device": Object {
    "context": Object {
      "id": "16613d16bf18d5674ad8c5d3ef071a3835ceb2b09fc3d5df592c0925f2e881f9",
    },
    "id": "4ed5ac9fcf3638b1ef16c7800a3f6d7c5b4aa092b9afe1dca139c65e97f023f1",
    "name": "Mensa Tablet",
    "owners": Array [
      Object {
        "id": "bb9b812794666578a86254afca49deef281fd5caf4566db9915cf4d5ecde53dd",
      },
    ],
  },
}
`;

exports[`Device Admin schould update device not owned by User [Mutation] 1`] = `
Object {
  "updateDevice": Object {
    "device": Object {
      "context": Object {
        "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
      },
      "id": "9db9921b0d8132795cb5e77a90572fce5e94b90a8f44bd48417dff8cf7b66834",
      "name": "RenamedTestDevice",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  },
}
`;

exports[`Device Admin schould update device owned by User [Mutation] 1`] = `
Object {
  "updateDevice": Object {
    "device": Object {
      "context": Object {
        "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
      },
      "id": "4ed5ac9fcf3638b1ef16c7800a3f6d7c5b4aa092b9afe1dca139c65e97f023f1",
      "name": "RenamedTestDevice",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  },
}
`;

exports[`Device Anonym should create a Device [Mutation] 1`] = `
Object {
  "context": null,
  "name": "TestDevice",
}
`;

exports[`Device User schould create device [Mutation] 1`] = `
Object {
  "context": null,
  "name": "TestDevice",
}
`;

exports[`Device User schould return all devices [Query] 1`] = `
Object {
  "devices": Array [
    Object {
      "context": Object {
        "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
      },
      "id": "13624f62a9b2a01b18d0ae062bf9ad978d0fe5265b9e059957f15c331ae11282",
      "name": "Forum Fernseher",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  ],
}
`;

exports[`Device User schould return device owned by User [Query] 1`] = `
Object {
  "device": Object {
    "context": Object {
      "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
    },
    "id": "13624f62a9b2a01b18d0ae062bf9ad978d0fe5265b9e059957f15c331ae11282",
    "name": "Forum Fernseher",
    "owners": Array [
      Object {
        "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
      },
    ],
  },
}
`;

exports[`Device User schould update device owned by User [Mutation] 1`] = `
Object {
  "updateDevice": Object {
    "device": Object {
      "context": Object {
        "id": "7df600774ceaa14488143c9d7877fd71662f4750c7c1c77aede7aa684d7c16f1",
      },
      "id": "13624f62a9b2a01b18d0ae062bf9ad978d0fe5265b9e059957f15c331ae11282",
      "name": "RenamedTestDevice",
      "owners": Array [
        Object {
          "id": "bb4028c0657dfd7092f316dcbf5b006dce87217141e1b974441f040126acc142",
        },
      ],
    },
  },
}
`;
